src/components/ElectionMap.js

```diff
@@ -303,10 +303,10 @@ class ElectionMap extends SvgChart {
     // return match && match.complete ? 0 : this.options().size
   }
 
-  opacity(d) {
-    return 1
-    const match = this.dataLookup[d.id]
-    return match && match.complete ? 1 : 0.5
+  opacity(data) {
+    const match = this.dataLookup[data.id]
+    return match && match.opacity != null ? match.opacity : 1
```

Comment:
The opacity computation should happen within this file, not the container. To separate visual encoding logic from data.

Reply:
Hmm… I have to disagree (quite strongly) on this one, because as the ElectionMap gets re-used to display multiple aspects of the election result, the code in this file will become increasingly complex, trying to account for all the visualizations.

e.g.

```js
  if (!match) {
    return 1
  }
  if (match.displayMode === 'heatmap') {
    if (match.areaStatus === 'winning') {
      return scaleLinear(/* ... */)
    } else if (match.areaStatus === 'sent') {
      return scaleLinear(/* ... */)
// (...so on...)
```

Not only the logic, but the interface would become more complex. for example, here’s what would happen to IMapZone to make election map have all the data necessary to compute the opacity:

```diff
 interface IMapZone {
   id: number
   partyId: number
   complete: boolean
   show: boolean
+  displayMode: 'heatmap' | 'winningCandidate'
+  areaStatus?: 'winning' | 'sent' | null
+  displayedCandidateScore?: number
+  winningCandidateScore?: number
+  totalVotersCount?: number
 }
```

Creating more visualizations would require editing code at 2 places — modifying the container to pass the required data to ElectionMap, and modifying the ElectionMap (which is 400+ lines already) to be able to make sense of that data. This shows a clear violation of the open-closed principle…

In fact, in my opinion, the ElectionMap would be easier to use and understand how did it become a more dumb component. i.e. it receives only what it needs to know to display it, and displays it, e.g.:

```diff
 interface IMapZone {
   id: number
-  partyId: number
-  complete: boolean
-  show: boolean
+  fill: 'solid' | 'striped'
+  color: string
+  opacity: number
 }
```

This would allow more visualization to be created just by creating a new container with little to no modification to the ElectionMap.

Reply:
fair enough. agree with passing the computed encoding as the logic for computation is too complex.